internal/gps/selection.go

```diff
@@ -59,13 +62,43 @@ func (s *selection) popSelection() (atomWithPackages, bool) {
 	return sel.a, sel.first
 }
 
+func (s *selection) findCaseConflicts(pr ProjectRoot) (bool, ProjectRoot) {
+	prlist, has := s.prLenMap[len(pr)]
+	if !has {
+		return false, ""
+	}
+
+	// TODO(sdboyer) bug here if it's possible that strings.ToLower() could
+	// change the length of the string
+	lowpr := strings.ToLower(string(pr))
+	for _, existing := range prlist {
+		if lowpr != strings.ToLower(string(existing)) {
+			continue
+		}
+		// If the converted strings match, then whatever we figure out here will
+		// be definitive - we needn't walk the rest of the slice.
+		if pr == existing {
+			return false, ""
+		} else {
+			return true, existing
+		}
+	}
+
+	return false, ""
+}
+
 func (s *selection) pushDep(dep dependency) {
-	s.deps[dep.dep.Ident.ProjectRoot] = append(s.deps[dep.dep.Ident.ProjectRoot], dep)
+	pr := dep.dep.Ident.ProjectRoot
+	s.deps[pr] = append(s.deps[pr], dep)
+	s.prLenMap[len(pr)] = append(s.prLenMap[len(pr)], pr)
 }
 
 func (s *selection) popDep(id ProjectIdentifier) (dep dependency) {
 	deps := s.deps[id.ProjectRoot]
 	dep, s.deps[id.ProjectRoot] = deps[len(deps)-1], deps[:len(deps)-1]
+
+	prlist := s.prLenMap[len(id.ProjectRoot)]
+	s.prLenMap[len(id.ProjectRoot)] = prlist[:len(prlist)-1]
```

Comment:
What is this supposed to do?

Reply:
is the question more about what `prLenMap` does in general, or just this specific segment here?

it's popping the last element off the slice of project roots on the `prLenMap`. it's dual to append we do up in `*selection.pushDep()`.

Reply:
Makes total sense now. Looks like I forgot Go slice semantics. 😛